### ISSUE TITLE
core-js issue after gridsome upgrade

### ISSUE THREAD

**Issue by oliverpool** [oliverfindl/vue-svg-inline-loader#39]
Thank for this cool package!

I just upgraded my gridsome app and it now fails to compile (using `yarn`) with the following errors
```
ERROR  Failed to compile with 67 errors                                                                                                                                                                   6:27:45 PM

These dependencies were not found:

* core-js/modules/es.array.concat in ./node_modules/gridsome/app/fetch.js, ./node_modules/gridsome/app/head.js and 2 others
* core-js/modules/es.array.for-each in ./node_modules/gridsome/app/head.js, ./node_modules/gridsome/app/directives/image.js and 2 others
* core-js/modules/es.array.from in ./node_modules/gridsome/app/directives/image.js, ./node_modules/cache-loader/dist/cjs.js??ref--1-0!./node_modules/babel-loader/lib??ref--1-1!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader/lib??vue-loader-options!./node_modules/vue-svg-inline-loader/dist/index.min.js??ref--0-2!./src/components/ScrollWords.vue?vue&type=script&lang=js&
* core-js/modules/es.array.index-of in ./node_modules/gridsome/app/utils/class.js
* core-js/modules/es.array.is-array in ./node_modules/gridsome/app/router.js, ./node_modules/gridsome/app/utils/class.js
* core-js/modules/es.array.join in ./node_modules/gridsome/app/components/Image.js, ./node_modules/gridsome/app/utils/helpers.js
* core-js/modules/es.array.map in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/router.js and 2 others
* core-js/modules/es.array.reduce in ./node_modules/cache-loader/dist/cjs.js??ref--1-0!./node_modules/babel-loader/lib??ref--1-1!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader/lib??vue-loader-options!./node_modules/vue-svg-inline-loader/dist/index.min.js??ref--0-2!./src/components/ScrollWords.vue?vue&type=script&lang=js&
* core-js/modules/es.array.slice in ./node_modules/gridsome/app/components/Image.js, ./node_modules/gridsome/app/utils/helpers.js and 1 other
* core-js/modules/es.date.to-string in ./node_modules/cache-loader/dist/cjs.js??ref--1-0!./node_modules/babel-loader/lib??ref--1-1!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader/lib??vue-loader-options!./node_modules/vue-svg-inline-loader/dist/index.min.js??ref--0-2!./src/layouts/Default.vue?vue&type=script&lang=js&
* core-js/modules/es.function.name in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/fetch.js and 1 other
* core-js/modules/es.number.constructor in ./node_modules/gridsome/app/components/Link.js, ./node_modules/cache-loader/dist/cjs.js??ref--1-0!./node_modules/babel-loader/lib??ref--1-1!./node_modules/cache-loader/dist/cjs.js??ref--0-0!./node_modules/vue-loader/lib??vue-loader-options!./node_modules/vue-svg-inline-loader/dist/index.min.js??ref--0-2!./src/components/ScrollWords.vue?vue&type=script&lang=js&
* core-js/modules/es.object.to-string in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/app.js and 4 others
* core-js/modules/es.regexp.constructor in ./node_modules/gridsome/app/components/Link.js, ./node_modules/gridsome/app/utils/helpers.js
* core-js/modules/es.regexp.exec in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/components/Link.js and 2 others
* core-js/modules/es.regexp.to-string in ./node_modules/gridsome/app/components/Link.js, ./node_modules/gridsome/app/utils/helpers.js
* core-js/modules/es.string.ends-with in ./node_modules/gridsome/app/directives/image.js
* core-js/modules/es.string.iterator in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/app.js and 2 others
* core-js/modules/es.string.link in ./node_modules/gridsome/app/head.js
* core-js/modules/es.string.replace in ./node_modules/gridsome/app/utils/helpers.js, ./node_modules/gridsome/app/utils/class.js
* core-js/modules/es.string.search in ./node_modules/gridsome/app/entry.client.js
* core-js/modules/es.string.split in ./node_modules/gridsome/app/utils/helpers.js, ./node_modules/gridsome/app/utils/class.js
* core-js/modules/es.string.starts-with in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/components/Link.js and 1 other
* core-js/modules/es.string.trim in ./node_modules/gridsome/app/utils/class.js
* core-js/modules/es.symbol in ./node_modules/gridsome/app/app.js
* core-js/modules/es.symbol.description in ./node_modules/gridsome/app/app.js
* core-js/modules/es.symbol.iterator in ./node_modules/gridsome/app/app.js
* core-js/modules/web.dom-collections.for-each in ./node_modules/gridsome/app/head.js, ./node_modules/gridsome/app/directives/image.js and 2 others
* core-js/modules/web.dom-collections.iterator in ./node_modules/gridsome/app/entry.client.js, ./node_modules/gridsome/app/app.js
```

If I completely remove `vue-svg-inline-loader`, everything works again.

`yarn list --depth 0`:
```
yarn list v1.22.0
,,,
├─ @babel/preset-env@7.8.4
├─ @babel/runtime@7.8.4
...
├─ @gridsome/cli@0.3.1
...
├─ core-js-compat@3.6.4
├─ core-js@2.6.11
├─ core-util-is@1.0.2
...
├─ vue-svg-inline-loader@1.4.5
...
├─ vue@2.6.11
```

I feel that the issue is due to the usage of the `2.*` version of `core-js` (wanted by this plugin) while gridsome 0.7 requires `"@vue/babel-preset-app@^4.2.2` which expects `core-js@^3.6.4`.

This is probably related to https://github.com/gridsome/gridsome/pull/982...

Is there any way to make this plugin use `core-js` 3 ? (or any better way to fix this?)

**Comment by oliverfindl** [comment]
Hello,

you can try to load non-transpilled version of this plugin, which is not using [core-js](https://github.com/zloirock/core-js) at all, although it needs to be installed. Look at [line 17](https://github.com/oliverfindl/vue-svg-inline-loader/blob/36fc9718f66665555c87b3e1295b242caea9bff4/examples/gridsome/gridsome.config.js#L17), but alter path to something like this `__dirname + "node_modules/vue-svg-inline-loader/src/index.js"` (not tested). Personally, I don't think it's [core-js](https://github.com/zloirock/core-js) related issue. Each package should have its own node_modules, specifically to adress this kind of conflicts.

If it breaks right after you install this plugin (like [this issue](https://github.com/oliverfindl/vue-svg-inline-loader/issues/37)), unfortunately I don't have solution for you.

Please let me know, if non-transpilled version works.

Thanks.

**Comment by oliverpool** [comment]
Hi,
thanks for you quick answer and suggestions.

Here is what I achieved after some testing:

## Downgrading gridsome :heavy_check_mark: 
Downgrading gridsome to version `0.7.12` fixes the issue 

## Changing the `loader` :stop_sign:
With gridsome 0.7.13, just changing the loader to `__dirname + "node_modules/vue-svg-inline-loader/src/index.js"` still made gridsome fail (same error as my first message)

## Removing `vue-svg-inline-loader` dependency :heavy_check_mark: 
And saving the `src/index.js` file to `src/external/vue-svg-inline-loader.js` and pointing the `loader` to it works fine! (it seems that gridsome already requires all the dependencies needed by the not transpiled version)

```js
const { join } = require('path');
/// ...
.loader(join(__dirname, 'src/external/vue-svg-inline-loader.js'))
```

**Comment by oliverfindl** [comment]
Hello,

well this looks strange. Similar as in [issue](https://github.com/oliverfindl/vue-svg-inline-loader/issues/37) mentioned above. Both this issues are related to [@vue/babel-preset-app](https://github.com/vuejs/vue-cli/tree/dev/packages/@vue/babel-preset-app). I have no clue, why is this happening and therefore, what needs to be fixed in this loader.

Upgrading [core-js](https://github.com/zloirock/core-js)@2 to [core-js](https://github.com/zloirock/core-js)@3 together with dumping deprecated [@babel/polyfill](https://github.com/babel/babel/tree/master/packages/babel-polyfill) will require some testing (more info [here](https://github.com/oliverfindl/vue-svg-inline-loader/issues/25)) in environments, that I don't have enough free time to setup and perform them.

If you or anyone else, who is reading this issue, have any idea how to solve this issue, feel free to comment here.

Thanks

**Comment by oliverpool** [comment]
Running `yarn add @babel/runtime` didn't solve the issue. But digging further (using `yarn why core-js`), I think that the issue lies in `yarn` "hoisting". 

It installed different versions of `core-js` : one being in the main `node_module` and the other being inside a nested `node_module`. Somehow `yarn` decides to put the version `2.*` at the root and `3.*`is nested.

I found another workaround to force the situation to be reversed:

## Adding `core-js` as a dependency :heavy_check_mark:

I ran `yarn add "core-js@^3.6.4` and forced the `resolutions`:
```
"resolutions": {
    "vue-svg-inline-loader/core-js": "^2.6.11",
    "vue-svg-inline-loader/@babel/polyfill/core-js": "^2.6.11",
    "core-js": "^3.6.4"
  }
```

And everything seems fine!

---

I guess the issue lies by `gridsome` and that it doesn't explicitly imports `core-js`. And then yarn takes some unfortunate decisions, when your library is used (i.e. putting the version 2.* of core-js at the root)

**Comment by oliverpool** [comment]
I am reporting this upstream (https://github.com/gridsome/gridsome/pull/994), since I think the issue does not lie in this package!

**Comment by oliverfindl** [comment]
Hello,

much thanks for taking your time performing this investigation!
I'm sure it will help other developers too!

Thanks

**Comment by oliverpool** [comment]
Thank your for your help and quick replies :smiley:

**Comment by afenton90** [comment]
I came across this issue today and found that including this package caused all of `core-js` to be imported regardless of whether I had pinned it at `3.6.*` and used the `src` version of this package.

Perhaps a re-think of how it is packaged with something like https://rollupjs.org/guide/en/ may be better

**Comment by oliverfindl** [comment]
Hello,

adding [Rollup](https://github.com/rollup/rollup) (or other bundler) in current situation won't solve this issue. To solve it, I should define [core-js](https://github.com/zloirock/core-js) as [dev dependency](https://docs.npmjs.com/files/package.json#devdependencies) and then use bundler on it. But at that moment, all dependencies could be defined as [dev dependencies](https://docs.npmjs.com/files/package.json#devdependencies) which is against npm [dependencies](https://docs.npmjs.com/files/package.json#dependencies) semantics, so I'm strictly against this kind of solution (or in my opinion - workaround).
Other solution would be, to remove transpilled version of this loader with all its dependencies, which would result in breaking changes for some users on older [Node.js](https://github.com/nodejs/node) environments. So this is also no way to go for me.

As mentioned [here](https://github.com/gridsome/gridsome/issues/1018), same (or similar) issue is occurring without this loader in place. So I can guess, it's issue related to [gridsome](https://github.com/gridsome/gridsome) itself and not issue of this loader. In addition, there were no issues reported outside of [gridsome](https://github.com/gridsome/gridsome) use case.

Please try to wait untill [gridsome](https://github.com/gridsome/gridsome) devs resolve it at their end by adding [core-js](https://github.com/zloirock/core-js) as [peer dependency](https://docs.npmjs.com/files/package.json#peerdependencies) or something similar.

Thanks

**Comment by oliverfindl** [comment]
Hello guys,

I had some spare time today and installed [gridsome](https://github.com/gridsome/gridsome)@0.7.13 (used [this](https://gridsome.org/starters/default-starter/) guide and checked version in `package-lock.json` after install) and [vue-svg-inline-loader](https://github.com/oliverfindl/vue-svg-inline-loader)@1.4.6 via npm. Used SVG image same way as in my example [here](https://github.com/oliverfindl/vue-svg-inline-loader/tree/master/examples/gridsome). I didn't encounter this issue nor while running dev server, nor while building production version.

Can somebody share dummy repo, where is this issue reproducible?

Thanks

**Comment by oliverpool** [comment]
Using `npm` seems to work fine.
The issue only appears when running with `yarn` (at least on my codebase)

**Comment by IlCallo** [comment]
Can confirm this happens with other framework too, I'm hitting this while testing new Quasar Framework `@quasar/app` version (not released yet).
In Quasar case, the core-js 3 dependency is in a sub-package, so your version is hoisted at root level instead.

@oliverpool workaround works, but I don't think it can be a way to deal with the problem in a definite way

**Comment by ststaynov** [comment]
Any potential timeline on the resolution of this problem? It's indeed not gridsome-specific

**Comment by oliverfindl** [comment]
Hello,

to be honest, I dont think this issue is caused by this package. This behavior is most probably result of some other issue from different package. Please read all comments here, especially [this comment](https://github.com/oliverfindl/vue-svg-inline-loader/issues/39#issuecomment-587727729) and also comments in [issue](https://github.com/oliverfindl/vue-svg-inline-loader/issues/37) mentioned in it. From my discoveries back then, this issue could be caused by missing peer dependency definition in some of dependencies in dependency tree. Its possible, that you will get same result with any other random library using [core-js](https://github.com/zloirock/core-js)@2.

I had plans for complete rewrite of this plugin, which would solve this issue, but unfortunately my time is very limited and I switched focus to [vue-svg-inline-plugin](https://github.com/oliverfindl/vue-svg-inline-plugin), which serves same purpose, but works on different level. So this package is now in kinda abandoned state (well, it still works for my projects and I dont have any issues using [core-js](https://github.com/zloirock/core-js)@3 alongside with this plugin, which use [core-js](https://github.com/zloirock/core-js)@2). My ultimate goal was to share some logic (e.g.: for merging node attributes) between these two packages, but I dont see this happening anytime soon, because it would be necessary to rewrite both of them.

If you are interested in hotfix, somebody already forked this package, upgraded it to [core-js](https://github.com/zloirock/core-js)@3 and published to npm: [vue-svg-inline-loader-corejs3](https://www.npmjs.com/package/vue-svg-inline-loader-corejs3), but according npm weekly download count (300 vs 9000), its not very popular.

If you have some free time, you can track down your dependencies and try to find one responsible for this behavior. From my initial research (half year ago), I suspected package [@vue/babel-preset-app](https://github.com/vuejs/vue-cli/tree/dev/packages/@vue/babel-preset-app) or some of its dependencies to cause this issue.

Please let me know if you have more questions or something isnt clear to you.

Thanks.

**Comment by lllopo** [comment]
@oliverfindl I strongly hope you do not abandon this. It is literally **the only** plugin I've found that works reliably and well in one simple task - inlining SVG's in html files of Webpack projects. I know it is not it's purpose, but it actually does it very well and works fine with Webpack 5. So, please ... please ... keep supporting it. It is a life saver. I can't imagine how there is no good plugin for this out there, but it isn't. This is the only one. You can even drop the "vue" part and just name it "svg-inline-loader" as it works just fine in many scenarios. And yes ... it needs upgrading of core-js version.

**Comment by oliverfindl** [comment]
Hello,

thanks for your kind words. :)

To keep my sanity about this ongoing "issue", I just released version 2.0.0 without [core-js](https://github.com/zloirock/core-js) altogether. This introduces breaking changes for old Node environments, which are not supported anymore (sorry in advance fellow developers). Rest of code is same as was in version 1.5.1, so no breaking changes here.

Originally, I had plans for revamping [loader options](https://github.com/oliverfindl/vue-svg-inline-loader#configuration), creating merge attributes function based on [Vue@3](https://github.com/vuejs/vue-next) logic and extract shared logic with [vue-svg-inline-plugin](https://github.com/oliverfindl/vue-svg-inline-plugin) to another package, but it has to wait for another major version release, which should be complete rewrite of this loader.

Closing this issue as resolved.

Thanks.
